rwengine/src/engine/Garage.cpp

```diff
@@ -0,0 +1,889 @@
+        auto doorModel = doorObject->getModelInfo<BaseModelInfo>();
+        auto collision = doorModel->getCollision();
+        // Original behavior - game actually subtracts 0.1f
+        doorHeight =
+            collision->boundingBox.max.z - collision->boundingBox.min.z - 0.1f;
+    }
+
+    if (secondDoorObject) {
+        startPositionSecondDoor = secondDoorObject->getPosition();
+    }
+
+    step /= doorHeight;
+
+    switch (type) {
+        case Garage::Type::Mission:
+        case Garage::Type::CollectCars1:
+        case Garage::Type::CollectCars2:
+        case Garage::Type::MissionForCarToComeOut:
+        case Garage::Type::MissionKeepCar:
+        case Garage::Type::Hideout1:
+        case Garage::Type::Hideout2:
+        case Garage::Type::Hideout3:
+        case Garage::Type::MissionToOpenAndClose:
+        case Garage::Type::MissionForSpecificCar:
+        case Garage::Type::MissionKeepCarAndRemainClosed: {
+            state = State::Closed;
+            break;
+        }
+
+        case Garage::Type::BombShop1:
+        case Garage::Type::BombShop2:
+        case Garage::Type::BombShop3:
+        case Garage::Type::Respray:
+        case Garage::Type::Crusher: {
+            state = State::Opened;
+            break;
+        }
+    }
+
+    if (state == State::Closed) {
+        fraction = 0.f;
+    } else {
+        fraction = 1.f;
+    }
+
+    if (doorObject) {
+        updateDoor();
+    }
+}
+
+Garage::~Garage() {
+}
+
+void Garage::makeDoorSwing() {
+    // This is permanent, you can't restore it
+    // back to non swing just like in original game
+    // Values are from original game
+    if (!swingType) {
+        swingType = true;
+        doorHeight /= 2.0f;
+        doorHeight -= 0.1f;
+    }
+}
+
+bool Garage::isTargetInsideGarage() {
+    return state == State::Closed && isObjectInsideGarage(target);
+}
+
+void Garage::activate() {
+    active = true;
+
+    if (type == Type::MissionForCarToComeOut && state == State::Closed) {
+        state = State::Opening;
+    }
+}
+
+void Garage::deactivate() {
+    active = false;
+}
+
+void Garage::open() {
+    if (state == State::Closed || state == State::Closing) {
+        state = State::Opening;
+    }
+}
+
+void Garage::close() {
+    if (state == State::Opened || state == State::Opening) {
+        state = State::Closing;
+    }
+}
+
+float Garage::getDistanceToGarage(glm::vec3 point) {
+    // Seems like original game ignores z axis
+    float dx = std::max({min.x - point.x, 0.f, point.x - max.x});
+    float dy = std::max({min.y - point.y, 0.f, point.y - max.y});
+
+    return std::sqrt(dx * dx + dy * dy);
+}
+
+bool Garage::isObjectInsideGarage(GameObject* object) {
+    auto p = object->getPosition();
+
+    // Do basic check first
+    if (p.x < min.x) return false;
+    if (p.y < min.y) return false;
+    if (p.z < min.z) return false;
+    if (p.x > max.x) return false;
+    if (p.y > max.y) return false;
+    if (p.z > max.z) return false;
```

Comment:
nit: this sort of check is a recurring pattern and I wonder if we should have a function for it (if there isn't one already somewhere)